# Ex1 - Filtering and Sorting Data

This time we are going to pull data directly from the internet.
Special thanks to: https://github.com/justmarkham for sharing the dataset and materials.

### Step 1. Import the necessary libraries

In [2]:
import pandas as pd

### Step 2. Import the dataset from this [address](https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv). 

In [3]:
file_path = r'https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv'

### Step 3. Assign it to a variable called chipo.

In [4]:
chipo = pd.read_csv('https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv', sep='\t')

In [5]:
chipo.head(10)

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   
5         3         1                           Chicken Bowl   
6         3         1                          Side of Chips   
7         4         1                          Steak Burrito   
8         4         1                       Steak Soft Tacos   
9         5         1                          Steak Burrito   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
5  [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...    $10.98   
6                                                NaN     $1.69   
7  [Tomatillo Red Chili Salsa, [Fajita Vegetables...    $11.75   
8  [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...     $9.25   
9  [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...     $9.25

### Step 4. How many products cost more than $10.00?

In order to be able to use comparison operators, we need to make sure the column item_price is of a numerical type.The item_column contains values in a string format using a dollar sign. We need to clean the data by removing the dollar sign and turn the string values to float.
1. Remove the Dollar Sign and Convert to Float:

In [19]:
chipo['item_price'] = chipo["item_price"].str.replace("$", "").astype(float)

AttributeError: Can only use .str accessor with string values!

Checking if there are missing values in the column : 

In [20]:
# Check for missing values in 'item_price' column
missing_values = chipo["item_price"].isnull() 

In [21]:
missing_values

0       False
1       False
2       False
3       False
4       False
        ...  
4617    False
4618    False
4619    False
4620    False
4621    False
Name: item_price, Length: 4622, dtype: bool

In [22]:
chipo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   order_id            4622 non-null   int64  
 1   quantity            4622 non-null   int64  
 2   item_name           4622 non-null   object 
 3   choice_description  3376 non-null   object 
 4   item_price          4622 non-null   float64
dtypes: float64(1), int64(2), object(2)
memory usage: 180.7+ KB


Length is 4622 and entries 4622, looks like there no missing Nan Values.

### Notes

- Step 2: Handle any potential missing values in item_price (if any).For example, we could fill missing values with 0 or drop them
- chipo['item_price'] = chipo['item_price'].fillna(0)  # if we wanted to fill with 0
- chipo = chipo.dropna(subset=['item_price'])  # if we wanted to drop rows with missing item_price


##### Now we can filter items that cost more than $10.00

Checking if we have successfuly removed $ sign and we are able to use comparison operator now  :

In [24]:
filtered_items = chipo[chipo['item_price'] > 10.00]

In [16]:
filtered_items

order_id  quantity           item_name  \
4            2         2        Chicken Bowl   
5            3         1        Chicken Bowl   
7            4         1       Steak Burrito   
13           7         1        Chicken Bowl   
23          12         1     Chicken Burrito   
...        ...       ...                 ...   
4610      1830         1       Steak Burrito   
4611      1830         1      Veggie Burrito   
4617      1833         1       Steak Burrito   
4618      1833         1       Steak Burrito   
4619      1834         1  Chicken Salad Bowl   

                                     choice_description  item_price  
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98  
5     [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...       10.98  
7     [Tomatillo Red Chili Salsa, [Fajita Vegetables...       11.75  
13    [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...       11.25  
23    [[Tomatillo-Green Chili Salsa (Medium), Tomati...       10.98  
...                                                 ...         ...  
4610  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...       11.75  
4611  [Tomatillo Green Chili Salsa, [Rice, Fajita Ve...       11.25  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...       11.75  
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...       11.75  
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...       11.25  

[1130 rows x 5 columns]

#### Notes :
- df['item_price'].str.replace('$', ''): This removes the dollar sign from each value in the item_price column. The str.replace() function is used perform string replacement.
- .astype(float): Converts the resulting string values to float. This allows for numerical comparisons.

### Step 5. What is the price of each item? 
###### print a data frame with only two columns item_name and item_price

### Step 6. Sort by the name of the item

### Step 7. What was the quantity of the most expensive item ordered?

### Step 8. How many times was a Veggie Salad Bowl ordered?

### Step 9. How many times did someone order more than one Canned Soda?